### PR TITLE
Follow-up of #5655: add mmult and Tmmult to petsc sparse matrix and petsc parallel sparse matrix

### LINKED ISSUE #5647: Need PETScWrappers::MPI::SparseMatrix::mmult that is similar to TrilinosWrappers::SparseMatrix::mmult and SparseMatrix::mmult
Both TrilinosWrappers::SparseMatrix and SparseMatrix have mmult operation that does C = A * diag(V) * B, but PETScWrappers::MPI::SparseMatrix does not. I think it would nice to have it. Besides I actually need it to compute the Schur complement of mass matrix: B*{diag(M)}^{-1}*B^T in my application.

I think the implementation can be like this:

1. make a copy of A, named A_copy
2. A_copy.MatDiagonalScale(NULL, V)
3. MatMatMult(A_copy, B, MAT_INITIAL_MATRIX, PETSC_DEFAULT, C)

### PR DESCRIPTION
**Pull request by Jie-Cheng**
Fixes #5647.

### CODE REVIEW

File: include/deal.II/lac/petsc_parallel_sparse_matrix.h [hunk 1]
```diff
@@ -408,6 +408,26 @@ namespace PETScWrappers
        */
       IndexSet locally_owned_range_indices() const;
 
+      /**
+       * Perform the matrix-matrix multiplication $C = AB$, or,
+       * $C = A \text{diag}(V) B$ given a compatible vector $V$.
+       *
+       * This function calls <code>MatrixBase::mmult</code> to do the actual work.
```
Comment (bangerth) [inline include/deal.II/lac/petsc_parallel_sparse_matrix.h]: If you write this as `MatrixBase::mmult()`, then you don't need the `<code>...</code>` markup -- doxygen will automatically link to that function.

Same below of course.

File: include/deal.II/lac/petsc_sparse_matrix.h [hunk 2]
```diff
@@ -209,6 +209,26 @@ namespace PETScWrappers
      */
     size_t n() const;
 
+    /**
+      * Perform the matrix-matrix multiplication $C = AB$, or,
+      * $C = A \text{diag}(V) B$ given a compatible vector $V$.
+      *
+      * This function calls <code>MatrixBase::mmult</code> to do the actual work.
```
Comment (bangerth) [inline include/deal.II/lac/petsc_sparse_matrix.h]: and here and below

**Comment by Jie-Cheng** [comment]
@bangerth Please let me know how to add tests for this, and more importantly, which cases should be tested.

File: include/deal.II/lac/petsc_parallel_sparse_matrix.h (hunk 1 above)
Comment (Jie-Cheng) [reply include/deal.II/lac/petsc_parallel_sparse_matrix.h]: Thanks, that will be fixed in the next round.

**Comment by bangerth** [comment]
Take a look here: http://dealii.org/developer/developers/testsuite.html . I think this explains in some detail how this is all supposed to work, but feel free to ask us questions! The test for your particular case should go into the `tests/petsc/` directory, where there are already many other PETSc tests.

**Comment by bangerth** [comment]
This is the follow-up to #5655, which addresses #5647.

**Comment by Jie-Cheng** [comment]
Added some tests for `PETScWrappers::SparseMatrix::mmult` and `PETScWrappers::SparseMatrix::Tmmult`, changed the vector type from `VectorBase` to `MPI::Vector`.
`PETScWrappers::MPI::SparseMatrix` was not tested because it uses the same underlying functions as `PETScWrappers::SparseMatrix`.  @bangerth  please take a look.

**Comment by davydden** [comment]
@bangerth since you originally reviewed the PR, would you please have another look?

**Comment by davydden** [comment]
./run-tests

File: source/lac/petsc_matrix_base.cc [hunk 3]
```diff
@@ -541,7 +541,7 @@ namespace PETScWrappers
           AssertThrow (ierr == 0, ExcPETScError(ierr));
           if (transpose_left)
             {
-              ierr = MatTranspose(tmp, MAT_REUSE_MATRIX, &tmp);
```
Comment (davydden) [inline source/lac/petsc_matrix_base.cc]: @Jie-Cheng one of the testers failed to build this PR:

```
[557/859] Building CXX object source/lac/CMakeFiles/obj_lac_debug.dir/sparse_matrix.cc.o
FAILED: /usr/bin/clang++   -DDEBUG -DTBB_DO_ASSERT=1 -DTBB_IMPLEMENT_CPP0X=1 -DTBB_USE_DEBUG -Isource/lac -Iinclude -I/home/bob/source/include -I/home/bob/source/bundled/tbb41_20130401oss/include -I/home/bob/source/bundled/umfpack/UMFPACK/Include -I/home/bob/source/bundled/umfpack/AMD/Include -I/home/bob/source/bundled/boost-1.62.0/include -I/home/bob/source/bundled/muparser_v2_2_4/include -I/usr/lib/openmpi/include -I/usr/lib/openmpi/include/openmpi -I/home/bob/libs/parmetis-4.0.3/include -I/home/bob/libs/trilinos-12.4.2-Source/include -I/home/bob/libs/petsc-3.6.4/include -I/home/bob/libs/hdf5-1.8.15-patch1/include -I/home/bob/libs/oce-OCE-0.17/include/oce -I/home/bob/libs/p4est-1.1/FAST/include -I/home/bob/libs/slepc-3.6.0/include    -pedantic -fPIC -Wall -Wextra -Wpointer-arith -Wwrite-strings -Wsynth -Wsign-compare -Wswitch -Woverloaded-virtual -Wno-deprecated-declarations -Qunused-arguments -Wno-unsupported-friend -Wno-unused-parameter -Wno-unused-function -std=c++11 -Wno-parentheses -O0 -ggdb -MMD -MT source/lac/CMakeFiles/obj_lac_debug.dir/petsc_matrix_base.cc.o -MF "source/lac/CMakeFiles/obj_lac_debug.dir/petsc_matrix_base.cc.o.d" -o source/lac/CMakeFiles/obj_lac_debug.dir/petsc_matrix_base.cc.o -c /home/bob/source/source/lac/petsc_matrix_base.cc
/home/bob/source/source/lac/petsc_matrix_base.cc:544:40: error: use of undeclared identifier 'MAT_INPLACE_MATRIX'
              ierr = MatTranspose(tmp, MAT_INPLACE_MATRIX, &tmp);
                                       ^
1 error generated.
ninja: build stopped: subcommand failed.
build FAILED
```

you probably need to guard this line into something like
```
#if DEAL_II_PETSC_VERSION_LT(3,5,0)
...
#else
..
#endif
```
depending on when `MatTranspose` changed.

Comment (Jie-Cheng) [reply source/lac/petsc_matrix_base.cc]: Thanks. `MatTranspose` was changed in PETSc 3.8. I'll update.

Comment (davydden) [reply source/lac/petsc_matrix_base.cc]: @Jie-Cheng did you have a chance to have a look at this?

**Comment by Jie-Cheng** [comment]
@davydden and @bangerth  Sorry for the extremely late update. I have been busy these days. I just added the version guard and tested it locally. Let's see how the tests go. Meanwhile, if you find anything wrong in the code, please let me know.

File: include/deal.II/lac/petsc_matrix_base.h [hunk 4]
```diff
@@ -975,7 +975,7 @@ namespace PETScWrappers
      */
     void mmult (MatrixBase       &C,
                 const MatrixBase &B,
-                const VectorBase &V = VectorBase()) const;
+                const VectorBase &V) const;
```
Comment (drwells) [inline include/deal.II/lac/petsc_matrix_base.h]: I know that I am late to the party, but I would like to check to make sure I understand what is going on here. This function won't work in the same way it used to if we require a vector here: what is wrong with leaving the default argument in?

Comment (Jie-Cheng) [reply include/deal.II/lac/petsc_matrix_base.h]: This function is implemented as a protected member in `MatrixBase` which will not be called directly. The `SparseMatrix::mmult` and `MPI::SparseMatrix::mmult` implement their own default arguments.

File: source/lac/petsc_parallel_sparse_matrix.cc [hunk 5]
```diff
@@ -707,6 +707,22 @@ namespace PETScWrappers
       return indices;
     }
 
+    void
+    SparseMatrix::mmult (SparseMatrix       &C,
+                         const SparseMatrix &B,
+                         const MPI::Vector  &V) const
+    {
+      MatrixBase::mmult (C, B, V);
```
Comment (bangerth) [inline source/lac/petsc_parallel_sparse_matrix.cc]: can you add (here and in all of the other functions you implement) a comment of the form
```
  // simply forward to the protected member function of the base class that
  // takes abstract matrix and vector arguments (to which the compiler automatically
  // casts the arguments)
```

Comment (Jie-Cheng) [reply source/lac/petsc_parallel_sparse_matrix.cc]: Thanks, just added this comment to `SparseMatrix::mmult`, `SparseMatrix::Tmmult`, `MPI::SparseMatrix::mmult` and `MPI::SparseMatrix::Tmmult` and rebased on the latest master.

**Comment by bangerth** [comment]
Great -- thank you for your patience!

OK to merge once the tester is happy.

**Pull request merged**